### ISSUE TITLE
Does the "clip gradient" configuration apply to the weights of all layers?

### ISSUE THREAD

**Issue by cuihenggang** [dmlc/mxnet#898]
I have some questions about the "clip gradient" configuration.

I remember in Caffe, when the "clip gradient" configuration is set for the SGD solver, it will possibly down scale the gradients, such that the *summation* of the L2 norms of the gradients of all the layers is smaller than the specified number (e.g., 10).

I'm wondering is the behavior the same in MXNet? I have this question because I think MXNet will send to other GPUs the weight updates of one layer as soon as the backward propagation of that layer is done, without waiting for the other layers. So I guess it's not possible for MXNet to clip the gradients based on the summation of the L2 norms of all weights?

**Comment by piiswrong** [comment]
It's possible. But it will slowdown performance. you need to write your own training loop for that

**Comment by cuihenggang** [comment]
@piiswrong Thank you for pointing out this, and actually I'm more curious about what's the current default behavior of gradient clipping in MXNet. Will it clip the gradient of each weight independently?

**Comment by piiswrong** [comment]
yes

**Comment by tqchen** [comment]
closing for now as it is answered
